pkg/controller/computenodeopenstack/computenodeopenstack_controller.go

```diff
@@ -82,26 +87,54 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		return err
 	}
 
-	err = c.Watch(&source.Kind{Type: &machinev1beta1.MachineSet{}}, &handler.EnqueueRequestForOwner{
+	err = c.Watch(&source.Kind{Type: &batchv1.Job{}}, &handler.EnqueueRequestForOwner{
 		IsController: true,
 		OwnerType:    &computenodev1alpha1.ComputeNodeOpenStack{},
 	})
 	if err != nil {
 		return err
 	}
 
-	err = c.Watch(&source.Kind{Type: &batchv1.Job{}}, &handler.EnqueueRequestForOwner{
+	// Watch ConfigMaps owned by ComputeNodeOpenStack
+	err = c.Watch(&source.Kind{Type: &corev1.ConfigMap{}}, &handler.EnqueueRequestForOwner{
 		IsController: true,
 		OwnerType:    &computenodev1alpha1.ComputeNodeOpenStack{},
 	})
 	if err != nil {
 		return err
 	}
 
-	// Watch ConfigMaps owned by ComputeNodeOpenStack
-	err = c.Watch(&source.Kind{Type: &corev1.ConfigMap{}}, &handler.EnqueueRequestForOwner{
-		IsController: true,
-		OwnerType:    &computenodev1alpha1.ComputeNodeOpenStack{},
+	// watch for MachineSet with compute-node-operator CR owner ref
+	ospMachineSetFn := handler.ToRequestsFunc(func(o handler.MapObject) []reconcile.Request {
+		cc := mgr.GetClient()
+		result := []reconcile.Request{}
+		ms := &machinev1beta1.MachineSet{}
+		key := client.ObjectKey{Namespace: o.Meta.GetNamespace(), Name: o.Meta.GetName()}
+		err := cc.Get(context.Background(), key, ms)
+		if err != nil && !errors.IsNotFound(err) {
+			log.Error(err, "Unable to retrieve MachineSet %v")
+			return nil
+		}
+
+		for label := range ms.ObjectMeta.GetLabels() {
+			if label == ownerLabelSelector {
+				log.Info(fmt.Sprintf("MachineSet object %s marked with OSP owner ref: %s", o.Meta.GetName(), label))
+				// return namespace and ms.Name
+				name := client.ObjectKey{
+					Namespace: ms.Namespace,
```

Comment:
we need to return the CR namespace and name, not the ones belonging to the machineset

Reply:
yes correct, done